Fixed UniteWithInputDirectory with relative or ~ paths

louispan opened the pull request:
#925 `UniteWithInputDirectory files_rec/async` only works with absolute paths. It doesn't work with `.` or `../foo` or `~/bar` paths.

Below is my changes that made it work. There may be a better way though

Shougo commented on autoload/vital/_unite/Prelude.vim:
```diff
@@ -275,7 +275,7 @@ function! s:substitute_path_separator(path) abort
 endfunction
 
 function! s:path2directory(path) abort
-  return s:substitute_path_separator(isdirectory(a:path) ? a:path : fnamemodify(a:path, ':p:h'))
```
Please don't change vital's library behavior.
It is overwritten when vital is updated.

Shougo replied:
You should create another function in unite/util.vim to convert the directory path.

louispan commented:
Thank you for your suggestions. I've reverted the vital changes and added a `unite#util#path2absolute_directory` function that is only used by `UniteWithInputDirectory`.
I've also added a feature so that you can specify `!` which gets resolved to the project directory.
I've also fixed #927 which adds a `-L` option to `file_rec/asyc` find so that it follows symbolic links to match the behaviour of `file` and `file_rec`, so users can switch between them transparently.

louispan commented:
I thought I didn't need to fix grep and find. It looks like `.` works but not `~`. I'll close this pull request until I've fixed it.

Shougo commented:
OK.

louispan commented:
Hi, I've made a bit more extensive changes, but now find, grep and UniteWithInputDirectory are used more consistently. I've also added extra source path arg, ? (for buffer directory) and ?! (for buffer project directory).

Shougo commented on autoload/unite/sources/rec.vim:
```diff
@@ -293,7 +293,8 @@ function! s:source_file_async.gather_candidates(args, context) "{{{
   " Note: If find command and args used, uses whole command line.
-  let commands = vimproc#parser#split_args(command) + paths
+  " -L follows symbolic links to have the same behaviour as file_rec
+  let commands = vimproc#parser#split_args(command) + ['-L'] + paths
```
`-L` is for find only option.
You must not add it by default.

Shougo commented on autoload/unite/sources/rec.vim:
```diff
@@ -293,7 +293,8 @@ function! s:source_file_async.gather_candidates(args, context) "{{{
+  let commands = vimproc#parser#split_args(command) + ['-L'] + paths
   if args[0] ==# 'find'
     " Default option.
     let commands += ['-path', '*/\.git/*', '-prune',
```
If you want to add `-L` by default, you should add the option in it.

louispan replied:
I think it should be there by default, otherwise file_rec/async will have different behaviour to file_rec. I think changing between file_rec and file_rec/async should to transparent to the user.

Shougo replied:
You should change `g:unite_source_rec_async_command` default value to `find -L`.

louispan replied:
Ah, I understand now. Good idea

Shougo commented:
Thanks.  I reviewed it.

louispan commented:
You may merge the UniteWithInputDirectory changes whenever you are ready :)

louispan commented:
I'm going to re-open this pull request to point to another branch, as I'll be making minor tweaks in my `master`.
